scheduler/feasible.go

```diff
@@ -768,3 +812,229 @@ OUTER:
+					available[d] = 0
+					desiredCount -= unused
+				}
+			}
+		}
+
+		// We couldn't match the request for the device
+		if desiredCount > 0 {
+			return false
+		}
+	}
+
+	// Only satisfied if there are no more devices to place
+	return true
+}
+
+// nodeDeviceMatches checks if the device matches the request and its
+// constraints. It doesn't check the count.
+func nodeDeviceMatches(ctx Context, d *structs.NodeDeviceResource, req *structs.RequestedDevice) bool {
+	if !d.ID().Matches(req.ID()) {
+		return false
+	}
+
+	// There are no constraints to consider
+	if len(req.Constraints) == 0 {
+		return true
+	}
+
+	for _, c := range req.Constraints {
+		// Resolve the targets
+		lVal, ok := resolveDeviceTarget(c.LTarget, d)
+		if !ok {
+			return false
+		}
+		rVal, ok := resolveDeviceTarget(c.RTarget, d)
+		if !ok {
+			return false
+		}
+
+		// Check if satisfied
+		if !checkAttributeConstraint(ctx, c.Operand, lVal, rVal) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// resolveDeviceTarget is used to resolve the LTarget and RTarget of a Constraint
+// when used on a device
+func resolveDeviceTarget(target string, d *structs.NodeDeviceResource) (*psstructs.Attribute, bool) {
+	// If no prefix, this must be a literal value
+	if !strings.HasPrefix(target, "${") {
+		return psstructs.ParseAttribute(target), true
+	}
+
+	// Handle the interpolations
+	switch {
+	case "${driver.model}" == target:
+		return psstructs.NewStringAttribute(d.Name), true
+
+	case "${driver.vendor}" == target:
+		return psstructs.NewStringAttribute(d.Vendor), true
+
+	case "${driver.type}" == target:
+		return psstructs.NewStringAttribute(d.Type), true
+
+	case strings.HasPrefix(target, "${driver.attr."):
+		attr := strings.TrimSuffix(strings.TrimPrefix(target, "${driver.attr."), "}")
+		val, ok := d.Attributes[attr]
+		return val, ok
+
+	default:
+		return nil, false
+	}
+}
+
+// checkAttributeConstraint checks if a constraint is satisfied
+func checkAttributeConstraint(ctx Context, operand string, lVal, rVal *psstructs.Attribute) bool {
+	// Check for constraints not handled by this checker.
+	switch operand {
+	case structs.ConstraintDistinctHosts, structs.ConstraintDistinctProperty:
+		return true
+	default:
+		break
+	}
+
+	switch operand {
+	case "<", "<=", ">", ">=", "=", "==", "is", "!=", "not":
+		v, ok := lVal.Compare(rVal)
+		if !ok {
+			return false
+		}
+
+		switch operand {
+		case "not", "!=":
+			return v != 0
+		case "is", "==", "=":
+			return v == 0
+		case "<":
+			return v == -1
+		case "<=":
+			return v != 1
+		case ">":
+			return v == 1
+		case ">=":
+			return v != -1
+		default:
+			return false
+		}
+
+	case structs.ConstraintVersion:
+		return checkAttributeVersionMatch(ctx, lVal, rVal)
+	case structs.ConstraintRegex:
+		ls, ok := lVal.GetString()
+		rs, ok2 := rVal.GetString()
+		if !ok || !ok2 {
+			return false
+		}
+		return checkRegexpMatch(ctx, ls, rs)
+	case structs.ConstraintSetContains:
```

Comment:
Should we also add support for ConstraintSetContainsAny here? (we've wanted to do so since affinities introduced it, but haven't gotten around to it, might as well do it now.)